docker role and playbook

NakulManchanda opened the pull request:
Signed-off-by: Nakul Manchanda <nakul.manchanda@ibm.com>

- PR has following changes:

    **Evironment Vars**
    - it introduce some new variables
       ZOWE_EXTERNAL_HOST - how to access zowe system externally
       ZOWE_ZOS_HOST - where is zos machine installed
       ZOWE_ZOSMF_HOST
       ZOWE_ZSS_HOST
      existing ZOWE_EXPLORER_HOST - this is different from all obove
    - It adds new test scope "docker test" to test docker image         
       
    **Playbook & Roles**   
    - new "docker" role
    - install-docker.yml - new docker install
    - add docker logs to show-logs.yml playbook

    **Sanity test** 
    - requires new environment variable ZOWE_EXTERNAL_HOST
    - eg docker - ZOWE_EXTERNAL_HOST=localhost, ZOWE_ZOS_HOST=marist-1, or any other target server
    - `npm run test:docker` - new npm script to run sanity test against local docker image
    - Sanity test refactor to replace SSH_HOST with ZOWE_EXTERNAL_HOST

      
    **Readme's**
     - updated docker readme
     - updated sanity test readme

<!-- Thank you for your pull request! Please provide a description of the changes in this PR in the field above and provide the following information. -->

Please check if your PR fulfills the following requirements. This is simply a reminder of what we are going to look for before merging your PR. If you don't know all of this information when you create this PR, don't worry. You can edit this template as you're working on it.

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Necessary documentation (if appropriate) have been added / updated
- [ ] DCO signoffs have been added to all commits, including this PR

#### PR type
What type of changes does your PR introduce to Zowe? _Put an `x` in the box that applies to this PR. If you're unsure about any of them, don't hesitate to ask._ 
- [ ] Bugfix <!-- non-breaking change which fixes an issue -->
- [ ] Feature <!-- non-breaking change which adds functionality -->
- [ ] Other... Please describe:

#### Relevant issues
<!-- Link to a relevant GitHub issue if any. If this PR applies to multiple issues, preface each issue reference with the "Fixes" keyword. For example, Fixes #123, Fixes #456. -->

Fixes <!-- Issue number -->

#### Changes proposed in this PR
<!-- Please describe your Pull Request. -->
- 

#### Does this PR introduce a breaking change?
<!-- Is this a fix or feature that would cause existing functionality to not work as expected? -->

- [ ] Yes
- [ ] No

<!-- If yes, please describe the impact and migration path below.-->

#### Does this PR do something the person installing Zowe should know about?

<!-- Is this a fix or a feature that changes customizable files (e.g. configuration files, sample JCL, etc),
product requirements, or other installation or configuration related area? If so, please fill out the template below. 
There is no need to report the need to restart the servers to activate the change. Note that the provided text will be formatted in 64-character lines, and that round brackets, ( and ), must always be used in matching pairs. -->

---
* Affected function: _general area of interest_ *

---
* Description: _1 line description_ *

---
* Part: _name of customizable file involved_  *

---
_multi-line description_

#### Is there a related doc issue or Pull Request? 
<!-- Link to a relevant documentation issue or Pull Request if any. -->

Doc issue/PR number: 

#### Other information

stevenhorsman commented on playbooks/install-docker.yml:
```diff
@@ -0,0 +1,90 @@
+    # Configure Zowe
+    - name: Reset zowe_launch_components for docker install
+      set_fact:
+        zowe_launch_components: app-server
```
should this be zss, or do you need to launch the app server on z/OS as well?

NakulManchanda replied:
modified

stevenhorsman commented on tests/sanity/README.md:
```diff
@@ -99,12 +99,14 @@ ZOWE_ROOT_DIR=/path/to/zowe \
   ZOSMF_PORT=10443 \
+  ZOSMF_HOST=test-server \
+  ZOWE_NONZOS_HOST= \
+  ZOWE_ZOS_HOST= \
```
Is ZOWE_ZOS_HOST used anywhere?

NakulManchanda replied:
removed - not used

NakulManchanda commented on playbooks/install-docker.yml:
```diff
@@ -0,0 +1,81 @@
+        name: zowe
+
+    # ============================================================================
+    # Configure Zowe
```
@jackjia-ibm one suggestion lets move this configure to docker role

NakulManchanda replied:
docker>
  tasks>
     configure-zowe-for-docker.yml

NakulManchanda replied:
add configure and uninstall to docker role

jackjia-ibm commented on playbooks/roles/configure/defaults/main.yml:
```diff
@@ -95,12 +98,18 @@ zowe_xmem_stc_group: ZWEADMIN
 # Restrict permissions to keystore/truststore after the certificate setup script
 zowe_lock_keystore: true
+#
+zowe_launch_components: api-mediation,files-api,jobs-api,explorer-jes,explorer-mvs,explorer-uss,zss,app-server
```
We had split api-mediation into 4 packages. Please update this default value to empty string. I believe this way, we can make it compatible with staging, and also won't conflict the value we may change in the future. For example, adding caching-service to the list.

NakulManchanda replied:
cleared value

jackjia-ibm commented on playbooks/roles/docker/defaults/main.yml:
```diff
@@ -0,0 +1,27 @@
+zowe_zss_https: False
+zowe_zss_port: 8542
+zowe_launch_components: zss
+zowe_sanity_test_testcases: '-- --config .mocharc-docker.yml'
```
I believe this value doesn't need to be here since the verify playbook/role won't read this config file.

NakulManchanda replied:
removed

jackjia-ibm commented on playbooks/roles/docker/tasks/docker-logs.yml:
```diff
@@ -0,0 +1,22 @@
+
+- name: get docker logs using docker ps
+  # raw: docker logs $(docker ps -a -q --filter ancestor={{zowe_docker_image}}:{{zowe_docker_tag}})
+  raw: docker logs $(docker ps -a -q 2>&1 | head -n 1)
```
This will show log of the first docker container. Maybe we can filter from the image name which should give you better chance to find the correct container.

BTW, with test run, do you find `container_id` could be empty?

jackjia-ibm replied:
Does `docker ps -a -q --filter ancestor=` give you failure? If so, you may also need to fix `playbooks/roles/docker/tasks/uninstall.yml` since it's using similar command.

NakulManchanda replied:
container id exists , we call that from  `install-docker.yml`

empty case - when we evoke it from `show-logs.yml`  as sometimes we quite before reaching install-docker log
,at that level we have zowe_docker_url only not even zowe_docker_image and zowe_image_tag
, so kept it without filters

NakulManchanda replied:
In pipeline, 

Called here:
https://github.com/zowe/zowe-install-packaging/blob/b01af5d55ed4030ebd6192fbb899e94d53b07b4f/tests/installation/src/__tests__/basic/install-docker.ts#L45

Evoked Here:
https://github.com/zowe/zowe-install-packaging/blob/b01af5d55ed4030ebd6192fbb899e94d53b07b4f/playbooks/show-logs.yml#L26

NakulManchanda replied:
modified as per query provided @jackjia-ibm

jackjia-ibm commented on playbooks/roles/docker/tasks/uninstall.yml:
```diff
@@ -0,0 +1,6 @@
+---
+- name: remove docker if running
+  raw: >-
+    docker rm -f $(docker ps -a -q --filter ancestor={{zowe_docker_image}}:{{zowe_docker_tag}})
```
Does this work well?

NakulManchanda replied:
yes, tried many times locally works great, as far as we are removing same image and tag

jackjia-ibm replied:
what happens in the case you mentioned in another block: at that level we have zowe_docker_url only not even zowe_docker_image and zowe_image_tag?

NakulManchanda replied:
it wont break pipeline `ignore_errors: True`
and help feature to re-run same playbook with same param locally

NakulManchanda replied:
we can always add additional ways to clear running container based on different conditionals and situation we encounter and enhance it

NakulManchanda replied:
modified as per query provided @jackjia-ibm

jackjia-ibm commented on playbooks/roles/start/tasks/wait_for_zowe.yml:
```diff
@@ -66,25 +70,31 @@
   delay: "{{ wait_for_zowe_service_delay | int }}"
+  # if zowe_zos_host and zowe_external_domain_name are not same, zlux could be
+  # started off z/OS and then we couldn't test right now
+  when: zowe_zos_host == zowe_external_domain_name and "api-mediation" in zowe_launch_components
```
Please change to `when: zowe_zos_host == zowe_external_domain_name and ("gateway" in zowe_launch_components or zowe_launch_components == "")`

jackjia-ibm commented on playbooks/roles/start/tasks/wait_for_zowe.yml:
```diff
@@ -99,6 +109,33 @@
   delay: "{{ wait_for_zowe_service_delay | int }}"
+  # if zowe_zos_host and zowe_external_domain_name are not same, zlux could be
+  # started off z/OS and then we couldn't test right now
+  when: zowe_zos_host == zowe_external_domain_name and "api-mediation" in zowe_launch_components
```
Please change to `when: zowe_zos_host == zowe_external_domain_name and ("api-catalog" in zowe_launch_components or zowe_launch_components == "")`

jackjia-ibm commented on playbooks/roles/start/tasks/wait_for_zowe.yml:
```diff
@@ -99,6 +109,33 @@
+  delay: "{{ wait_for_zowe_service_delay | int }}"
+  # if zowe_zos_host and zowe_external_domain_name are not same, zlux could be
+  # started off z/OS and then we couldn't test right now
+  when: zowe_zos_host == zowe_external_domain_name and "app-server" in zowe_launch_components
```
Please change to `when: zowe_zos_host == zowe_external_domain_name and ("app-server" in zowe_launch_components or zowe_launch_components == "")`

jackjia-ibm commented on playbooks/roles/start/tasks/wait_for_zowe.yml:
```diff
@@ -66,25 +70,31 @@
   delay: "{{ wait_for_zowe_service_delay | int }}"
+  # if zowe_zos_host and zowe_external_domain_name are not same, zlux could be
+  # started off z/OS and then we couldn't test right now
+  when: zowe_zos_host == zowe_external_domain_name and "jobs-api" in zowe_launch_components
```
Please change to `when: zowe_zos_host == zowe_external_domain_name and ("jobs-api" in zowe_launch_components or zowe_launch_components == "")`

jackjia-ibm commented on playbooks/roles/start/tasks/wait_for_zowe.yml:
```diff
@@ -42,21 +43,24 @@
   delay: "{{ wait_for_zowe_service_delay | int }}"
+  # if zowe_zos_host and zowe_external_domain_name are not same, zlux could be
+  # started off z/OS and then we couldn't test right now
+  when: zowe_zos_host == zowe_external_domain_name and "app-server" in zowe_launch_components
```
Please change to `when: zowe_zos_host == zowe_external_domain_name and ("app-server" in zowe_launch_components or zowe_launch_components == "")`

jackjia-ibm commented on tests/installation/src/constants.ts:
```diff
@@ -30,6 +30,10 @@ export const SANITY_TEST_REPORTS_DIR: string = path.resolve(SANITY_TEST_ROOT_DIR
 export const TEST_TIMEOUT_CONVENIENCE_BUILD: number = 60 * 60 * 1000;
+
+// 30 minutes timeout for install docker build
+export const TEST_TIMEOUT_DOCKER_BUILD: number = 30 * 60 * 1000;
```
I think this value is not used any more? Or do you plan to use it in `tests/installation/src/__tests__/basic/install-docker.ts`?

NakulManchanda replied:
no plans, i can remove it

jackjia-ibm reviewed (changes requested):
Thanks Nakul. There are few changes mostly related to the `zowe_launch_components` value

jackjia-ibm commented:
Please note, after you merge https://github.com/zowe/zowe-dockerfiles/pull/21, `Dockerfile.jenkins` will be removed. You will also need to change `Jenkinsfile` `docker build -f Dockerfile.jenkins` commands to use correct Dockerfile.

NakulManchanda commented:
@stevenhorsman @jackjia-ibm 
- changes look good now, lets merge this tomm,
- got thorough review from @jackjia-ibm - addressed all feedback based on alternatives provided

- it passed smpe successfully
- 160 pass 1 fail in convenience
- docker expected failure as image don't have zss and caching service changes

Overall, seems ready to merge, after last review and changes from jack

jackjia-ibm reviewed (approved):
Looks good to me. Thanks a lot Nakul!
